### DISCUSSION TITLE
Located new method returning consumed slice

### DISCUSSION THREAD

**Discussion by xkikeg** [winnow-rs/winnow#557] [Ideas]
Since `Located` holds `initial` internally, it'll be trivial to give "already consumed slices".
Currently I have a implementation to get consumed slices for line number offset by holding both `&str initial` and `Located<&str> input`, but if I can get `input.consumed()` via `Located`, that'll be easy to use. WDYT?

**Comment by epage** [comment]
If I'm understanding correctly, during your parse you are wanting to turn `Located::location` into `(line_num, col_num)`?  Is there a reason you are wanting to track that during a parse, rather than `Location::location`?

For myself, I've found byte offsets should exclusively be used, leaving `(line_num, col_num)` to only be calculated when displaying to users
- `col_num` can mean many things and, when rendering,. is dependent on what you are rendering to
- Bytes are simpler to use programmatically
- Byte tracking is faster

**Comment by xkikeg** [comment]
Thanks for the reply! Exactly true, I don't need line_number until I get an error (either parse error or semantic error).

What I want to achieve is similar to https://docs.rs/winnow/latest/winnow/_tutorial/chapter_7/index.html,
but with setting `annotate_snippets::Snippets::line_start` to avoid copying entire input `&str` for no reasons.

At that point, essentially I'll compute `1 + (initial...input).filter(|b| == b'\n').count()` as the `line_start`.

**Reply by epage** [reply]
Personally, I'd recommend sending the entire `&str` to annotate-snippets and using `fold(true)`.  This is what we are doing inside of cargo.  Generally errors like this are in a path where performance doesn't matter as much.

But that also gives me better context to understanding the problem.

You should be able to do something like
```rust
fn to_app_error(err: ParseError<&str, ContextError>) -> AppError {
    let input = err.input();
    let consumed = &input[0..err.offset()];
    let line_start_offset = consumed.rfind('\n').map(|offset| offset + 1).unwrap_or(0);
    let line_end_offset = input[line_start_offset..].find('\n').unwrap_or(input.len());
    let line = &input[line_start_offset..line_end_offset];
    let line_start = input[0..line_start_offset).lines();
    AppError::new(line, line_start, err.inner())
}
```

**Comment by xkikeg** [comment]
Thank you for suggestion!
entire copy might be still worth, I already allocated them as a `String` so it's not that bad to replicate them.

I'm not using `parse` / `ParseError` though, relying on `parse_next` so that I can drop the parsed result almost immediately.
So at the time of error, input is already moved to the beginning of the current chunk, not the original 0.

https://github.com/xkikeg/okane/blob/3aac9529e66a72aa27fdc9169041f729c83edde0/core/src/parse.rs#L118
anyway I already have the working code and it wouldn't be bad to have `initial` on my side. (Maybe I can use `CheckPoint` to be generalized)
so it's not critical to have the API specific to `Located`..
